### PR TITLE
feat: internal pre and post creation and post authentication triggers

### LINKED ISSUE #4908: Action Possibility on Local Users
As a developer I want to extend the workflow for local users that authenticate in ZITADEL so I am able to extend my register and login flow for my custom needs.

**Acceptance criteria**
- [x] I want to be able to add an action on local authentication
- [x] I want to be able to trigger the action Post Authentication
- [x] I want to be able to trigger the action Pre Creation
- [x] I want to be able to trigger the action  Post Creation
- [x] I want to be able to create the same actions as we have on external authentication

### LINKED ISSUE #5061: E2E tests in machines.cy.ts and humans.cy.ts are flaky
As a developer, I want the e2e tests for user specs to be deterministic so that I'm confident that the tests are meaningful.

**Acceptance criteria**
- [ ] tests in humans.cy.ts are not skipped
- [ ] tests in machines.cy.ts are not skipped
- [ ] running the Pipeline four times in a row succeeds regarding humans.cy.ts
- [ ] running the Pipeline four times in a row succeeds regarding machines.cy.ts

### PR DESCRIPTION
**Pull request by eliobischof**
closes #5061

- [x] Also adds fields ctx and authrequest to external flows ctx.v1 and mgmt to api.v1  

Open since first review:

- [x] Make sure scripts can't overwrite context.Context fields
- [ ] Ensure that validations execute. I added an e2e test that should test that validations execute. The tested action doesn't have the setUserMetadata available on the ctx.v1.mgmt object. I didn't find out the reason. Other actions that use this method pass. When this is fixed and the test still fails because validations don't run, we probably have to find another way of providing the methods. I'd still suggest to keep them as aligned as possible with the API for keeping the users and developers mental overhead as low as possible.
- [ ] There are some scary ZITADEL logs when the tests run that say that a policy is not found. Not sure if it's relevant.

Also refactors the e2e suite:

- [x] Prepares for #4532 
- [x] Spec files are run in isolated organizations
- [x] Redesigns API calls and awaiting eventual consistency
- [x] `npm run e2e` defaults to chrome that also enables testing u2f and passwordless when on the same origin using the cypress-grep plugin
- [x] Fixes user tests that succeeded even though they should have failed
- [x] Reenables chrome web security
- [x] Awaits actionable elements to be visible where possible
- [x] Makes login flow more idempotent
- [x] Caches sessions across specs
- [x] Fixes e2e prepare container bug

Common ACs

- [x] I am happy with the code
- [x] Short description of the feature/issue is added in the pr description
- [x] PR is linked to the corresponding user story
- [x] Acceptance criteria are met
- [x] All open todos and follow ups are defined in a new ticket and justified
- [x] Deviations from the acceptance criteria and design are agreed with the PO and documented.
- [x] No debug or dead code
- [x] Critical parts are tested automatically
- [x] Where possible E2E tests are implemented
- [x] Documentation/examples are up-to-date
- [x] All non-functional requirements are met
- [ ] Functionality of the acceptance criteria is checked manually on the dev system.

### CODE REVIEW

File: internal/api/ui/login/register_handler.go [hunk 1]
```diff
@@ -82,11 +82,48 @@ func (l *Login) handleRegisterCheck(w http.ResponseWriter, r *http.Request) {
...
-	user, err := l.command.RegisterHuman(setContext(r.Context(), resourceOwner), resourceOwner, data.toHumanDomain(), nil, nil, initCodeGenerator, emailCodeGenerator, phoneCodeGenerator)
+
+	// For consistency with the external authentication flow,
+	// the setMetadata() function is provided on the pre creation hook, for now,
+	// like for the ExternalAuthentication flow.
+	// If there is a need for additional context after registration,
+	// we could provide that method in the PostCreation trigger too,
+	// without breaking existing actions.
+	// Also, if that field is needed, we probably also should provide it
+	// for ExternalAuthentication.
```
Comment (eliobischof) [inline internal/api/ui/login/register_handler.go:93]: We should remove this inconsistent overhead code and just pass the mgmt server and auth request anyway

File: console/src/app/pages/users/user-detail/auth-user-detail/auth-user-mfa/auth-user-mfa.component.ts [hunk 2]
```diff
@@ -71,7 +71,7 @@ export class AuthUserMfaComponent implements OnInit, OnDestroy {
 
         const index = list.findIndex((mfa) => mfa.otp);
         if (index === -1) {
-          this.otpAvailable = true;
+          this.otpDisabled$.next(false);
```
Comment (eliobischof) [inline console/src/app/pages/users/user-detail/auth-user-detail/auth-user-mfa/auth-user-mfa.component.ts:74]: This nurns the OTP button to being enabled when it's already rendered

File: e2e/cypress/e2e/permissions/permissions.cy.ts [hunk 3]
```diff
@@ -1,44 +1,57 @@
...
-import { apiAuth } from '../../support/api/apiauth';
-import { ensureProjectExists, ensureProjectResourceDoesntExist, Roles } from '../../support/api/projects';
+import { ensureProjectExists } from 'support/api/projects';
+import { ensureRoleDoesntExist } from 'support/api/roles';
+import { newTarget } from 'support/api/target';
+import { ensureHumanDoesntExist, ensureHumanExists } from 'support/api/users';
+import { ZITADELTarget } from 'support/commands';
+import { loginname } from 'support/login/login';
+
+// TODO: Unskip when nil pointer bug is fixed for members api
```
Comment (eliobischof) [inline e2e/cypress/e2e/permissions/permissions.cy.ts:16]: @adlerhurst can you do that, plz?

**Comment by eliobischof** [comment]
@adlerhurst can you please help me with the linting errors? I can't reproduce them locally.

**Comment by adlerhurst** [comment]
> @adlerhurst can you please help me with the linting errors? I can't reproduce them locally.

Im not aware of goland. The errors are because of wrong formatted imports. You should see the same errors if you run it locally using `golangci-lint run ./...`

File: e2e/cypress/e2e/permissions/permissions.cy.ts (hunk 3 above)
Comment (adlerhurst) [reply e2e/cypress/e2e/permissions/permissions.cy.ts:16]: when do you get the panic?

Comment (adlerhurst) [reply e2e/cypress/e2e/permissions/permissions.cy.ts:16]: ah found it

**Comment by eliobischof** [comment]
> > @adlerhurst can you please help me with the linting errors? I can't reproduce them locally.
> 
> Im not aware of goland. The errors are because of wrong formatted imports. You should see the same errors if you run it locally using `golangci-lint run ./...`

Yeah, it don't.

image

File: e2e/cypress/e2e/permissions/permissions.cy.ts (hunk 3 above)
Comment (adlerhurst) [reply e2e/cypress/e2e/permissions/permissions.cy.ts:16]: @eliobischof I created a PR for this one. please merge by yourself

File: internal/api/ui/login/register_handler.go (hunk 1 above)
Comment (eliobischof) [reply internal/api/ui/login/register_handler.go:93]: let's deprecate the specific metadata code @adlerhurst?

**Comment by eliobischof** [comment]
@adlerhurst ctx.v1.ctx has neighter writable nor readable fields. Both (sub-)context fields key and value are unexported.

Example for the instance ID (note the unexported val field):
image

**Pull request closed without merging**